File: finalyearproject/repos/dkd-fraud/src/fraud/events.py
"""The three fraud Published-Language events (frozen registry, producer 10, key DID).
Payloads carry canonical IDs only — never PII (R6). Anything else is an R6 violation."""

from __future__ import annotations

import json
from typing import Any

from psycopg import Connection

from fraud import ids

TOPIC_SIGNAL_RAISED = "fraud.enforcement.FraudSignalRaised.v1"
TOPIC_ACCOUNT_HELD = "fraud.enforcement.AccountHeld.v1"
TOPIC_HOLD_RELEASED = "fraud.enforcement.AccountHoldReleased.v1"

PRODUCED_TOPICS = frozenset({TOPIC_SIGNAL_RAISED, TOPIC_ACCOUNT_HELD, TOPIC_HOLD_RELEASED})

CONSUMED_TOPICS = (
    "identity.party.KYCApproved.v1",
    "identity.party.KYCTierChanged.v1",
    "identity.party.PartySuspended.v1",
    "b2c.order.OrderPlaced.v1",
    "b2b.tradeorder.TradeOrderCreated.v1",
)


def _emit(cx: Connection[Any], topic: str, key: str, fields: dict[str, Any], now: int) -> str:
    if topic not in PRODUCED_TOPICS:
        raise ValueError(f"R6 violation: fraud may not produce {topic}")
    event_id = ids.new_event_id()
    payload: dict[str, Any] = {"eventId": event_id, "occurredAt": now, **fields}
    cx.execute(
        "INSERT INTO outbox(event_id, topic, partition_key, payload, occurred_at) "
        "VALUES (%s,%s,%s,%s::jsonb,%s)",
        (event_id, topic, key, json.dumps(payload), now),
    )
    return event_id


def fraud_signal_raised(
    cx: Connection[Any], subject_did: str, reason: str, risk_score: float, now: int
) -> str:
    return _emit(
        cx,
        TOPIC_SIGNAL_RAISED,
        subject_did,
        {"subjectDid": subject_did, "reason": reason, "riskScore": risk_score, "raisedAt": now},
        now,
    )


def account_held(
    cx: Connection[Any], subject_did: str, approver1: str, approver2: str, now: int
) -> str:
    return _emit(
        cx,
        TOPIC_ACCOUNT_HELD,
        subject_did,
        {"subjectDid": subject_did, "approver1": approver1, "approver2": approver2, "heldAt": now},
        now,
    )


def account_hold_released(cx: Connection[Any], subject_did: str, now: int) -> str:
    return _emit(
        cx,
        TOPIC_HOLD_RELEASED,
        subject_did,
        {"subjectDid": subject_did, "releasedAt": now},
        now,
    )

File: finalyearproject/repos/dkd-b2c-order/src/app/events.ts
// b2c Published-Language events into the transactional outbox. Topic names come from the
// vendored SDK registry (producer must be context 6 — R6); payloads carry IDs only, never
// PII (deliveryAddress stays in the DB, FR-MKT-004).
import { KafkaTopics as TOPICS, topicMeta } from "@dokandar/platform-sdk";
import type { PgTx } from "../persistence/pg.js";
import { OutboxStore } from "../store/spine.js";
import { newEventId } from "../domain/ids.js";
import type { OrderItem } from "../domain/order.js";
import { stringifyWithBigInt } from "../domain/json.js";

const B2C_CONTEXT = 6;

export class EventFactory {
  constructor(private readonly outbox: OutboxStore) {}

  async orderPlaced(tx: PgTx, ord: string, buyerDid: string, sellerDid: string,
                    items: OrderItem[], totalAmountPoisha: bigint, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_PLACED_V1, ord, {
      ord, buyerDid, sellerDid,
      items: items.map((it) => ({
        lineId: it.lineId, gpid: it.gpid, ppid: it.ppid ?? undefined,
        quantity: it.quantity, unit: it.unit,
        unitPricePoisha: it.unitPricePoisha,
      })),
      totalAmountPoisha,
      // finance's consumer resolves the escrow amount from amountPoisha-family fields
      amountPoisha: totalAmountPoisha,
      placedAt: now,
    }, now);
  }

  async paymentConfirmed(tx: PgTx, ord: string, escrowId: string, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_PAYMENT_CONFIRMED_V1, ord,
      { ord, escrowId, confirmedAt: now }, now);
  }

  async processingStarted(tx: PgTx, ord: string, sellerDid: string, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_PROCESSING_STARTED_V1, ord,
      { ord, sellerDid, startedAt: now }, now);
  }

  async orderShipped(tx: PgTx, ord: string, shipmentId: string, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_SHIPPED_V1, ord,
      { ord, shipmentId, shippedAt: now }, now);
  }

  async orderDelivered(tx: PgTx, ord: string, deliveredAt: number, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_DELIVERED_V1, ord,
      { ord, orderId: ord, deliveredAt }, now);
  }

  async orderCancelled(tx: PgTx, ord: string, reason: string, cancelledBy: string, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_CANCELLED_V1, ord,
      { ord, orderId: ord, reason, cancelledAt: now, cancelledBy }, now);
  }

  async orderRefunded(tx: PgTx, ord: string, refundAmountPoisha: bigint, now: number): Promise<void> {
    await this.emit(tx, TOPICS.B2C_ORDER_ORDER_REFUNDED_V1, ord,
      { ord, refundAmountPoisha, refundedAt: now }, now);
  }

  private async emit(tx: PgTx, topic: string, key: string,
                     fields: Record<string, unknown>, now: number): Promise<void> {
    if (topicMeta(topic).producer !== B2C_CONTEXT)
      throw new Error(`R6 violation: b2c may not produce ${topic}`);
    const eventId = newEventId();
    // stringifyWithBigInt renders poisha as exact integer literals on the Published-Language wire
    // (CC-CONS-03) — never a Number() float64 downcast.
    const payload = stringifyWithBigInt({ eventId, occurredAt: now, ...fields });
    await this.outbox.insert(tx, eventId, topic, key, payload, now);
  }
}
